tapir/coop/services/ResignMemberService.py

```diff
@@ -0,0 +1,84 @@
+import datetime
+
+from tapir.accounts.models import TapirUser
+from tapir.coop.models import ResignedMembership, ShareOwnership
+from tapir.coop.forms import MembershipCancelForm
+from tapir.shifts.models import (
+    ShiftAttendanceTemplate,
+    ShiftAttendance,
+    ShiftAccountEntry,
+)
+from tapir.shifts.services.frozen_status_service import FrozenStatusService
+from tapir.shifts.models import ShiftUserData, ShiftAttendanceMode
+
+from tapir.utils.shortcuts import get_timezone_aware_datetime
+
+
+class ResignMemberService:
+    @staticmethod
+    def update_shifts_and_shares(
+        form: MembershipCancelForm, member: ResignedMembership
+    ):
+        tapir_user: TapirUser = getattr(member.share_owner, "user", None)
+        if not tapir_user:
+            print("Couldn't find an existing Tapir user.")
+            return
+        shares = ShareOwnership.objects.filter(share_owner=member.share_owner)
+        if (
+            member.willing_to_gift_shares_to_coop
+            or member.transfering_shares_to is not None
+        ):
+            for share in shares:
+                if share.is_fully_paid:
+                    if member.transfering_shares_to is not None:
+                        ShareOwnership.objects.create(
+                            share_owner=member.transfering_shares_to,
+                            amount_paid=share.amount_paid,
+                            start_date=member.cancellation_date,
+                        )
+                        share.delete()
+                    else:
+                        share.is_active = False
+                else:
+                    raise ValueError(
+                        "Shareowner didn't fully paid all shares, so it cannot be tranfered to someone else."
+                    )
+                    break
+
+            for attendance_template in ShiftAttendanceTemplate.objects.filter(
+                user=tapir_user
+            ):
+                start_date = get_timezone_aware_datetime(
+                    member.cancellation_date, datetime.time()
+                )
+                attendance_template.cancel_attendances(start_date)
+                attendance_template.delete()
+
+                attendances = ShiftAttendance.objects.filter(
+                    user=tapir_user,
+                    slot__shift__start_time__gte=start_date,
+                    state=ShiftAttendance.State.PENDING,
+                )
+
+                for attendance in attendances:
+                    attendance.state = ShiftAttendance.State.CANCELLED
+                    attendance.save()
+
+                for shift_account_entry in ShiftAccountEntry.objects.filter(
+                    user=tapir_user
+                ):
+                    shift_account_entry.delete()
```

Comment:
Do we need to delete the shift account history of every member that leaves? We could just keep it.

Reply:
Deleting the `ShiftAccountEntry` won't prevent the counting from happening, it will just set their score to 0.

I think the problem is more that inactive members still get minus points at the start of each shift cycle.
If a member resigned immediately, then the ShareOwneship will have it's date end set and the member will be inactive, right?
So we could update tapir.shifts.services.shift_cycle_service.ShiftCycleService to check only active members instead of all members.

Reply:
I'll do the update on ShiftCycleService since it's not directly related to your changes

Reply:
I just checked the `ShiftCycleService` and the `ShiftExpectationService`, they already check if the member is active before removing points, so I guess the members that complained about receiving warning emails were still active?
Do you have case we could check?